nats.go

```diff
@@ -3613,6 +3620,35 @@ func (nc *Conn) unsubscribe(sub *Subscription, max int, drainMode bool) error {
 	return nil
 }
 
+// processControlFlow checks whether it is a control flow message and
+// handles it automatically.
+func (s *Subscription) processControlFlow(msg *Msg) (*Msg, error) {
+	fc := len(msg.Data) == 0 && msg.Header.Get(statusHdr) == flowControl
+	if !fc {
+		return msg, nil
+	}
+
+	// Return in case of an i/o or connection error on publish.
+	err := msg.Respond(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Peek at the channel and return the next one already
+	// in case there is one waiting.
+	s.mu.Lock()
```

Comment:
RLock?

Reply:
Updated to avoid taking the lock to get the sub.mch channel since already had a [snapshot copy of it from earlier](https://github.com/nats-io/nats.go/blob/5f6f07b870cf9652b7453f831ed83014eeaac411/nats.go#L3667-L3668) and passing it instead to the function as a receive only channel.